src/main/java/net/snowflake/ingest/streaming/internal/SubscopedTokenExternalVolumeManager.java

```diff
@@ -0,0 +1,230 @@
+  private static final Logging logger = new Logging(SubscopedTokenExternalVolumeManager.class);
+  // Reference to the external volume per table
+  private final Map<String, InternalStage> externalVolumeMap;
+
+  /** Increasing counter to generate a unique blob name */
+  private final AtomicLong counter;
+
+  // name of the owning client
+  private final String clientName;
+
+  private final String role;
+
+  // Reference to the Snowflake service client used for configure calls
+  private final SnowflakeServiceClient serviceClient;
+
+  // Client prefix generated by the Snowflake server
+  private final String clientPrefix;
+
+  // Deployment ID returned by the Snowflake server
+  private final Long deploymentId;
+
+  // concurrency control to avoid creating multiple ExternalVolume objects for the same table (if
+  // openChannel is called
+  // multiple times concurrently)
+  private final Object registerTableLock = new Object();
+
+  /**
+   * Constructor for ExternalVolumeManager
+   *
+   * @param role the role of the client
+   * @param clientName the name of the client
+   * @param snowflakeServiceClient the Snowflake service client used for configure calls
+   */
+  SubscopedTokenExternalVolumeManager(
+      String role, String clientName, SnowflakeServiceClient snowflakeServiceClient) {
+    this.clientName = clientName;
+    this.role = role;
+    this.counter = new AtomicLong(0);
+    this.serviceClient = snowflakeServiceClient;
+    this.externalVolumeMap = new ConcurrentHashMap<>();
+    try {
+      ClientConfigureResponse response =
+          this.serviceClient.clientConfigure(new ClientConfigureRequest(role));
+      this.clientPrefix = response.getClientPrefix();
+      this.deploymentId = response.getDeploymentId();
+    } catch (IngestResponseException | IOException e) {
+      throw new SFException(e, ErrorCode.CLIENT_CONFIGURE_FAILURE, e.getMessage());
+    }
+    logger.logDebug(
+        "Created SubscopedTokenExternalVolumeManager with clientName=%s and clientPrefix=%s",
+        clientName, clientPrefix);
+  }
+
+  /**
+   * Given a fully qualified table name, return the target storage by looking up the table name
+   *
+   * @param fullyQualifiedTableName the target fully qualified table name
+   * @return target storage
+   */
+  @Override
+  public InternalStage getStorage(String fullyQualifiedTableName) {
+    // Only one chunk per blob in Iceberg mode.
+    return getVolumeSafe(fullyQualifiedTableName);
+  }
+
+  /** Informs the storage manager about a new table that's being ingested into by the client. */
+  @Override
+  public void registerTable(TableRef tableRef) {
+    if (this.externalVolumeMap.containsKey(tableRef.fullyQualifiedName)) {
+      logger.logInfo(
+          "Skip registering table since its already been registered with the VolumeManager."
+              + " tableRef=%s",
+          tableRef);
+      return;
+    }
+
+    // future enhancement - per table locks instead of per-client lock
```

Comment:
Since you are using a ConcurrentHashMap don't you get per key locks for free anyways if you use `computeIfAbsent` etc.?

Reply:
That's correct, i somehow had the assumption that computeIfAbsent does not guarantee singular invocation of the function passed in, just read docs to confirm it does guarantee that. Switched to computeIfAbsent!